solana/accounting-contract/programs/accounting-contract/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+    use super::*;
+
+    pub fn set_admin(ctx: Context<SetAdmin>, new_admin: Pubkey) -> Result<()> {
+        let upgrade_authority = match ctx.accounts.program_account.upgrade_authority_address {
+            Some(address) => address,
+            None => return Err(ErrorCode::UpgradeAuthorityNotFound.into()),
+        };
+        require!(
+            ctx.accounts.caller.key() == upgrade_authority,
+            ErrorCode::Unauthorized
+        );
+        let admin = &mut ctx.accounts.admin_account.admin;
+        *admin = new_admin;
+        Ok(())
+    }
+
+    pub fn set_worker(ctx: Context<SetWorker>, new_worker: Pubkey) -> Result<()> {
+        require!(
+            ctx.accounts.admin_account.admin == ctx.accounts.caller.key(),
+            ErrorCode::Unauthorized
+        );
+        ctx.accounts.worker_account.worker = new_worker;
+        Ok(())
+    }
+
+    pub fn deposit_funds(ctx: Context<DepositFunds>, amount: u64) -> Result<()> {
+        system_program::transfer(
+            CpiContext::new(
+                ctx.accounts.system_program.to_account_info(),
+                system_program::Transfer {
+                    from: ctx.accounts.depositor.to_account_info(),
+                    to: ctx.accounts.treasury.to_account_info(),
+                },
+            ),
+            amount,
+        )?;
+        Ok(())
+    }
+
+    pub fn withdraw_funds(ctx: Context<WithdrawFunds>, amount: u64) -> Result<()> {
+        require!(
+            ctx.accounts.admin_account.admin == ctx.accounts.signer.key(),
+            ErrorCode::Unauthorized
+        );
+
+        require!(
+            ctx.accounts.treasury.to_account_info().lamports() >= amount,
+            ErrorCode::OutOfBalance
+        );
+
+        **ctx
+            .accounts
+            .treasury
+            .to_account_info()
+            .try_borrow_mut_lamports()? -= amount;
+        **ctx
+            .accounts
+            .beneficiary
+            .to_account_info()
+            .try_borrow_mut_lamports()? += amount;
+
+        Ok(())
+    }
+
+    pub fn create_pay_request(
+        ctx: Context<CreatePayRequest>,
+        nonce: u64,
+        amount: u64,
+    ) -> Result<()> {
+        require!(
+            ctx.accounts.worker.worker == ctx.accounts.signer.key(),
+            ErrorCode::Unauthorized
+        );
+
+        require!(
+            ctx.accounts.treasury.to_account_info().lamports() >= amount,
+            ErrorCode::OutOfBalance
+        );
+
+        require!(
+            ctx.accounts.account_nonce.nonce < nonce,
+            ErrorCode::InvalidNonce
+        );
+
+        **ctx
+            .accounts
+            .treasury
+            .to_account_info()
+            .try_borrow_mut_lamports()? -= amount;
+        **ctx
+            .accounts
+            .beneficiary
+            .to_account_info()
+            .try_borrow_mut_lamports()? += amount;
+
+        let payout_request = &mut ctx.accounts.pay_out_request;
```

Comment:
Also a dumb question: why do we need to mutate the payout request? 🤔

Reply:
1. Why do we need to mutate the payout request? 
So `Payout_Request` is a storage item that I'm maintaining to store the `beneficiary` , `nonce` and `amount`, so when someone submits a payout_request, this storage item gets created and updated. 
2. Why do we need the PayoutRequest account?
While you can track individual transactions via the explorer, i felt it was a bit tedious and a storage item could be easier since there was a lot of emphasis on having the ability to audit these cross-chain transactions, If we feel we don't need this storage item then we can remove it.

Reply:
I see, but wouldn't this storage only keep data of the latest transaction? There are a few ways to audit the transactions in solana, and it's way simpler that Ethereum because you don't have to go block by block. A simple way is to use [getSignaturesForAddress](https://solana.com/docs/rpc/http/getsignaturesforaddress) in combination with [getTransaction](https://solana.com/docs/rpc/http/gettransaction)

Reply:
> I see, but wouldn't this storage only keep data of the latest transaction? 

No it doesn't store only the latest transaction, it's address is derived with the beneficiary and nonce, It's sort of like double mapping keys, which is unique for each payout request. 
https://github.com/litentry/heima/blob/109ae6e206ec654a76d7848a7b38679fb848e3f2/solana/accounting-contract/programs/accounting-contract/src/lib.rs#L149

Reply:
> > A simple way is to use [getSignaturesForAddress](https://solana.com/docs/rpc/http/getsignaturesforaddress) in combination with [getTransaction](https://solana.com/docs/rpc/http/gettransaction)
> 
> 
> 
> This still requires filtering for the particular transaction and fetching the details for it, It still requires an upper and lower bound for the blocks. 
> 
> 
> 
> We are basically checking which all contexts this address is a part of and then fetching the details of those transactions, correct, wdyt? 
> 
> We can remove it if we feel it's not necessary

I think it's fine to keep it